pkg/util/http.go

```diff
@@ -374,3 +392,49 @@ func copyValues(src url.Values) url.Values {
 func IsHTTPStatusCode(code codes.Code) bool {
 	return int(code) >= 100 && int(code) < 600
 }
+
+// BufferProvider provides request body and decoding buffers.
+type BufferProvider struct {
+	size     int
+	body     *bytes.Buffer
+	decoding *bytes.Buffer
+}
+
+// NewBufferProvider returns a new BufferProvider given an initial buffer size.
+func NewBufferProvider(size int) BufferProvider {
+	return BufferProvider{
+		size: size,
+	}
+}
+
+// BodyBuffer gets the buffer for reading the request body.
+func (p *BufferProvider) BodyBuffer() *bytes.Buffer {
+	if p.body == nil {
+		p.body = bytes.NewBuffer(make([]byte, 0, p.size))
+	}
+	return p.body
+}
+
+// DecodingBuffer gets the buffer for decoding the request body.
+func (p *BufferProvider) DecodingBuffer() *bytes.Buffer {
+	if p.decoding == nil {
+		p.decoding = bytes.NewBuffer(make([]byte, 0, p.size))
+	}
+	return p.decoding
+}
+
+// ReplaceDecodingBuffer replaces the decoding buffer with a new one backed by buf.
+func (p *BufferProvider) ReplaceDecodingBuffer(buf []byte) {
+	p.decoding = bytes.NewBuffer(buf)
+}
+
+// CleanUp releases p to the pool.
+func (p *BufferProvider) CleanUp(pool *sync.Pool) {
```

Comment:
This should just be `Reset` and return to the pool be done outside of this method. (No need to make `BufferProvider` to depend on `sync.Pool`).

Reply:
> No need to make BufferProvider to depend on sync.Pool

Isn't this advice in contradiction to your suggestion of the `requestBuffers` struct though, which has a `Pool` pointer?

Reply:
> > No need to make BufferProvider to depend on sync.Pool
> 
> Isn't this advice in contradiction to your suggestion of the `requestBuffers` struct though, which has a `Pool` pointer?

Suggested `requestBuffers` is doing both: getting buffers from the pool, and putting them back to pool. `BufferProvider` in this only doing the latter.

Reply:
I see what you mean @pstibrany - I think I will check with you offline later, to fully understand :)

Reply:
Thanks for clarifying :)